File: cpp/Environment.cpp
#include "Environment.hpp"

namespace reactnativecss {
    namespace env {

        static std::shared_ptr<reactnativecss::Observable<double>> &widthRef() {
            static auto inst = reactnativecss::Observable<double>::create(0.0);
            return inst;
        }

        static std::shared_ptr<reactnativecss::Observable<double>> &heightRef() {
            static auto inst = reactnativecss::Observable<double>::create(0.0);
            return inst;
        }

        static std::shared_ptr<reactnativecss::Observable<double>> &scaleRef() {
            static auto inst = reactnativecss::Observable<double>::create(0.0);
            return inst;
        }

        static std::shared_ptr<reactnativecss::Observable<double>> &fontScaleRef() {
            static auto inst = reactnativecss::Observable<double>::create(0.0);
            return inst;
        }

        reactnativecss::Observable<double> &windowWidth() { return *widthRef(); }

        reactnativecss::Observable<double> &windowHeight() { return *heightRef(); }

        reactnativecss::Observable<double> &windowScale() { return *scaleRef(); }

        reactnativecss::Observable<double> &windowFontScale() { return *fontScaleRef(); }

        void setWindowDimensions(double width, double height, double scale, double fontScale) {
            widthRef()->set(width);
            heightRef()->set(height);
            scaleRef()->set(scale);
            fontScaleRef()->set(fontScale);
        }

    } // namespace env
} // namespace reactnativecss

File: src/native/js-lib/Environment.ts
/**
 * Environment: Window dimensions tracking
 * Converted from cpp/Environment.cpp
 */

import { Observable } from "./Observable";

const EnvironmentState = (() => {
  const widthObs = Observable.create(0);
  const heightObs = Observable.create(0);
  const scaleObs = Observable.create(0);
  const fontScaleObs = Observable.create(0);

  return {
    windowWidth(): Observable<number> {
      return widthObs;
    },

    windowHeight(): Observable<number> {
      return heightObs;
    },

    windowScale(): Observable<number> {
      return scaleObs;
    },

    windowFontScale(): Observable<number> {
      return fontScaleObs;
    },

    setWindowDimensions(
      width: number,
      height: number,
      scale: number,
      fontScale: number,
    ): void {
      widthObs.set(width);
      heightObs.set(height);
      scaleObs.set(scale);
      fontScaleObs.set(fontScale);
    },
  };
})();

export const env = EnvironmentState;
